### PR TITLE
S3 Bucket Policy: String Policy Input

### LINKED ISSUE #455: S3 BucketPolicy not able to put multiple StringEquals conditions
### What happened?

While creating s3.aws.crossplane.io/v1alpha2 BucketPolicy conditions for StringEquals. Seem I can only put 1 key-value pair. As for certain scenario, multi key-value pairs are needed

### How can we reproduce it?

Sample manifest:

```
apiVersion: s3.aws.crossplane.io/v1alpha2
kind: BucketPolicy
metadata:
  name: mybucketname.bucketpolicy
spec:
  forProvider:
    version: "2012-10-17"
    bucketName: mybucketname
    region: us-west-1
    statements:
      - sid: DownloadandUpload
        action:
          - "s3:GetObject"
          - "s3:GetObjectAcl"
          - "s3:GetObjectVersion"
          - "s3:PutObject"
          - "s3:PutObjectAcl"
          - "s3:DeleteObject"
          - "s3:DeleteObjectVersion"
        effect: Allow
        principal:
          allowAnon: true
        resource: 
          - "mybucketname"
          - "mybucketname/*"  
        condition:
          StringEquals:
            - key: "aws:Key1"
              stringValue: "value1"
            - key: "aws:Key2"
              stringValue: "value2"
            - key: "aws:Key3"
              stringValue: "value3"
            - key: "aws:Key4"
              stringValue: "value4"
  providerConfigRef:
    name: aws-provider
  deletionPolicy: Delete
```

### What environment did it happen in?
Crossplane version: v0.14.0
Provider-aws version: v0.15.0

### PR DESCRIPTION
**Pull request by krishchow**
### Description of your changes

Fixes #455

This PR aims to add an optional JSON input to the s3 bucket policy, this would allow users to choose between using an existing stringified s3 bucket policy, or defining a bucket policy in their manifest.

I have:

- [X] Read and followed Crossplane's [contribution process].
- [X] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Updated unit tests, manual testing

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/s3/v1alpha2/bucketpolicy_types.go [hunk 1]
```diff
@@ -28,16 +28,15 @@ type BucketPolicyParameters struct {
...
-	// This is the current IAM policy version
-	Version string `json:"version"`
-
-	// This is the policy's optional identifier
-	// +immutable
+	// JSONStatement is a stringified version of the S3 Bucket Policy.
+	// either jsonBody or policyBody must be specified in the policy
 	// +optional
-	ID string `json:"id,omitempty"`
+	JSONBody *string `json:"jsonBody,omitempty"`
```
Comment (muvaf) [inline apis/s3/v1alpha2/bucketpolicy_types.go:34]: What do you think about `rawPolicy` and `policy` field names? I feel like `Body` could be interpreted as only the `Statements` object in the JSON that doesn't include `Version`, but we need `Version` in this string.

**Review by muvaf — COMMENTED** [review]
Looking pretty good, I have only one comment regarding field names.

**Comment by muvaf** [comment]
@krishchow how do you feel about getting this merged before the feature freeze?

**Comment by krishchow** [comment]
Sound reasonable! I'll rebase and aim to make the relevant changes by EOD

File: apis/s3/v1alpha2/bucketpolicy_types.go (hunk 1 above)
Comment (krishchow) [reply apis/s3/v1alpha2/bucketpolicy_types.go]: Should be done now! :)

File: pkg/controller/s3/bucketpolicy/bucketpolicy.go [hunk 2]
```diff
@@ -44,6 +44,7 @@ const (
 	errDelete           = "failed to delete the policy for bucket"
 	errGet              = "failed to get BucketPolicy for bucket with name"
 	errUpdate           = "failed to update the policy for bucket"
+	errNotSpecified     = "failed to format bucketPolicy, no policyBody or jsonBody specified"
```
Comment (mcavoyk) [inline pkg/controller/s3/bucketpolicy/bucketpolicy.go:47]: Error string should be updated to reflect updated fields `rawPolicy` and `policy`

File: apis/s3/v1alpha3/bucketpolicy_types.go [hunk 3]
```diff
@@ -28,16 +28,15 @@ type BucketPolicyParameters struct {
...
 	// +optional
-	ID string `json:"id,omitempty"`
+	RawPolicy *string `json:"rawPolicy,omitempty"`
 
-	// This is the list of statement this policy applies
-	Statements []BucketPolicyStatement `json:"statements"`
+	// Policy is a well defined type which can be parsed into an JSON S3 Bucket Policy
+	// either jsonBody or policyBody must be specified in the policy
+	// +optional
+	Policy *BucketPolicyBody `json:"policy,omitempty"`
```
Comment (hasheddan) [inline apis/s3/v1alpha3/bucketpolicy_types.go:39]: This is a breaking API change here so we likely need to bump this to `v1alpha4`

**Review by hasheddan — CHANGES_REQUESTED** [review]

File: examples/s3/bucketpolicy-json.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+apiVersion: s3.aws.crossplane.io/v1alpha2
+kind: BucketPolicy
+metadata:
+  name: bucketpolicy-public
+spec:
+  forProvider:
+    region: us-west-1
+    bucketNameRef:
+      name: test-bucket
+    jsonBody: '{"Statement":[{"Action":"s3:ListBucket","Effect":"Allow","Principal":"*","Resource":"arn:aws:s3:::test-bucket"}],"Version":"2012-10-17"}'
```
Comment (mcavoyk) [inline examples/s3/bucketpolicy-json.yaml:10]: Example should be updated to reflect newer `rawPolicy` field.

Comment (mcavoyk) [reply examples/s3/bucketpolicy-json.yaml:10]: Also nit, but I would prefer the example `rawPolicy` have more readable format than compact json, like

```yaml
rawPolicy: |
  {
   "Statement":[
      {
         "Action":"s3:ListBucket",
         "Effect":"Allow",
         "Principal":"*",
         "Resource":"arn:aws:s3:::test-bucket"
      }
   ],
   "Version":"2012-10-17"

```

**Review by hasheddan — APPROVED** [review]
@krishchow mentioned to me offline that the bump to `v1alpha3` happened since the most recent release (`v0.16.0`), so this API has already been versioned correctly with the breaking changes (see https://doc.crds.dev/github.com/crossplane/provider-aws/s3.aws.crossplane.io/BucketPolicy/v1alpha2@v0.16.0)

File: apis/s3/v1alpha3/bucketpolicy_types.go [hunk 5]
```diff
@@ -53,6 +52,22 @@ type BucketPolicyParameters struct {
 	BucketNameSelector *xpv1.Selector `json:"bucketNameSelector,omitempty"`
 }
 
+// BucketPolicyBody represents an S3 bucket policy in the manifest
+type BucketPolicyBody struct {
+	// Version is the current IAM policy version
+	Version string `json:"version"`
```
Comment (mcavoyk) [inline apis/s3/v1alpha3/bucketpolicy_types.go:58]: Should this be an enum for either `2012-10-17` or `2008-10-17`. Those are the only two valid options I see from this IAM policy doc https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_version.html (not sure if s3 policy is different). 

Secondary question, should default to `2012-10-17`, unless this is being late initialized with AWS default?

File: apis/s3/v1alpha3/bucketpolicy_types.go [hunk 6]
```diff
@@ -28,16 +28,15 @@ type BucketPolicyParameters struct {
 	// +immutable
 	Region string `json:"region"`
 
-	// This is the current IAM policy version
-	Version string `json:"version"`
-
-	// This is the policy's optional identifier
-	// +immutable
+	// RawPolicy is a stringified version of the S3 Bucket Policy.
+	// either jsonBody or policyBody must be specified in the policy
```
Comment (muvaf) [inline apis/s3/v1alpha3/bucketpolicy_types.go:32]:
```suggestion
	// either rawPolicy or policy must be specified in the policy
```

**Review by muvaf — APPROVED** [review]
There are a few small comments, but I think it's almost ready. Let me know when it's ready to be merged. Thank you @krishchow !

File: examples/s3/bucketpolicy-json.yaml (hunk 4 above)
Comment (krishchow) [reply examples/s3/bucketpolicy-json.yaml]: Thanks for the review! These changes have been made and pushed :)

**Comment by krishchow** [comment]
Should be good to merge @muvaf

**Comment by muvaf** [comment]
@krishchow thank you!

**Pull request merged**